### PR TITLE
Fix git hooks prompt 

### LINKED ISSUE #26129: Git Hooks (pre-receive) with fail check drops Gitea to 502 Error
### Description

I have faced a situation where Git Hook drops Gitea to 502 Error.
Git Hook checks commits for some naming templates and raise error if check is not completed. All commits/PRs are blocked if check is failing.

Issue:
Gitea executes Git Hook on PR with non 0 error code. After receiving exit code 1 it drops to 502 Error

Expected behavior:
Show page with git console and error code from git. Do not drop to 502. 

How to reproduce:
1. Enable git hooks on Gitea and Configure Git Hook (pre-recive) on server side with this code:
```
#!/bin/bash

# Error message for BRANCH POLICY
error_msg_branch=$(cat <<-END

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@                                                                               @   
@   !!! Push not allowed by BRANCH NAME policy !!!                              @
@                                                                               @  
@   You branch should be named with these templaes:                             @
@     - master                                                                  @
@     - develop                                                                 @
@     - release/RC_*                                                            @
@     - feature/JIRA_ID-000--*                                                  @
@     - bugfix/*                                                                @
@     - hotfix/*                                                                @
@                                                                               @
@   Example: feature/MEC-167--add_new_functionaliry                             @
@                                                                               @
@   Wiki: https://confluence.ncloudtech.ru/display/KDM/Git+branch+naming        @                                                                        @
@                                                                               @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    
END
)

error_msg_commit=$(cat <<-END

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@                                                                                 @
@   !!! Your Commit does not have a referance to JIRA ticket !!!                  @
@                                                                                 @
@   Please correct your Git messages, and push again.                             @
@   Example: "[KDM-1234] This is a correct JIRA reference"                        @
@                                                                                 @
@   Wiki: https://confluence.ncloudtech.ru/pages/viewpage.action?pageId=201708163 @                                                                                @
@                                                                                 @
@   To FORCE push, use "bugfix" or "hotfix" in your commit message                @                                                                  @
@                                                                                 @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    
END
)



while read oldrev newrev refname
do
  #
  # Step 0:
  # Get Git commit information
  #
  BRANCH_NAME_FULL=$refname
  BRANCH_NAME="$(echo $BRANCH_NAME_FULL | sed 's/refs\/heads\///g')"
  COMMIT_MESSAGE=$(git log --format=%B -n 1)
  echo "[INFO] BRANCH NAME: $BRANCH_NAME"

  #
  # Step 1:
  # Policy - Check that brunch name is enforced by branch name
  #

  echo "[INFO] Policy - Check that brunch name is enforced by branch name"
  # Regexp for allowed names of branches
  branch_name_format='^master|^main|^develop|^release\/RC_.*|^feature\/[a-zA-Z0-9,\.\_\-]+-[0-9]+.*|^hotfix\/.*|^bugfix\/.*'
  
  if [[ ! $BRANCH_NAME =~ $branch_name_format ]]; then
      echo "$error_msg_branch" >&2
      exit 1
  else
      echo "Push is successful"
  fi

  #
  # Step 2:
  # Policy - Check commit message for JIRA issue number
  #

  # Configuration
  echo "[INFO] Policy - Check commit message for JIRA issue number"
  jiraIdRegex="[a-zA-Z0-9,\.\_\-]+-[0-9]+"
  fixMsgRegex="bugfix|hotfix"
  info_msg="[INFO] The commit message looks good"
  error_msg="[POLICY] The commit doesn't reference a JIRA issue"

  # Get all commits from this push
  for sha1Commit in $(git rev-list $oldrev..$newrev);
  do
    # Receive git commits sha from git history in chronologic order
    echo "[INFO] Processing commit with sha: $sha1Commit";
    # Get commit message from commit
    commitMessage=$(git log --format=%B -n 1 $sha1Commit)
    # Check with RegEX if commit has JIRA reference
    jiraIds=$(echo $commitMessage | grep -Eo $jiraIdRegex)
    fixMsg=$(echo $commitMessage | grep -Eo $fixMsgRegex)
    # Check if this commit urgent e.g. hotfox or bugfix
    if [[ -n "${fixMsg}" ]]; then
      echo "[WARNING] Found "bugfix|hotfix" in msg. Force skipping check for JiraID"
      exit 0
    fi
    # Check for JiraIDs in commit message
    echo "[INFO] Found JIRA IDs in commit: $jiraIds"
    if [[ -z "${jiraIds}" ]]; then
      echo "$error_msg: $commitMessage" >&2
      echo "$error_msg_commit" >&2
      exit 1
    fi
  done
done

#
# Exit
#
exit 0
```
2. Create new branch `feature/jira-123_some_text` and push to gitea
3. Create PR from `feature/jira-123_some_text` --> master
4. Get Error 502 on Gitea. The reason of this error, bacause git hook check failes. It not allows refs with PRs and exit with error 1. 
I expect, that it will show `$error_msg_branch >&2` from bash script in hook redirected to git log (How it is done on client side, if hook failes), but not prop to 502 Error

If i change   `branch_name_format='^master|^main|^develop|^release\/RC_.*|^feature\/[a-zA-Z0-9,\.\_\-]+-[0-9]+.*|^hotfix\/.*|^bugfix\/.*'` to `  branch_name_format='pull|^master|^main|^develop|^release\/RC_.*|^feature\/[a-zA-Z0-9,\.\_\-]+-[0-9]+.*|^hotfix\/.*|^bugfix\/.*' the check will not fail and PR working
`

### Gitea Version

1.20.0

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

Screenshot1
Screenshot2

### Git Version

_No response_

### Operating System

linux

### How are you running Gitea?

docker

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by CaiCandong**
close #26129

Before:

After:

### CODE REVIEW

**Comment by CaiCandong** [comment]
There is another bug here that can lead to the generation of dirty data: since Git operations rely on the creation of database PRs, when a Git operation fails, it's unable to roll back the changes in the database, resulting in the generation of dirty data.

Anyone have a good idea how to fix this bug?

- Screenshot 
gif
- Related code
https://github.com/go-gitea/gitea/blob/983167cf49cd64c7a1f28c39aff03e00dd387990/services/pull/pull.go#L53-L75

**Comment by CaiCandong** [comment]
> Anyone have a good idea how to fix this bug?

The bug will be fixed #26259

**Review by delvh — APPROVED** [review]

**Comment by wxiaoguang** [comment]
I do not think the UI is right.


image

**Comment by CaiCandong** [comment]
> I do not think the UI is right.

The prompt message is from a user-set hook, which isn't very pretty, but giving a prompt is a little friendlier than the previous 500.

**Comment by wxiaoguang** [comment]
But there are full of HTML tags. Are these tags from the git hook?

**Comment by CaiCandong** [comment]
> But there are full of HTML tags. Are these tags from the git hook?

HTML tags are not from a git hook, now I get what you're saying.

File: routers/web/repo/pull.go [hunk 1]
```diff
@@ -1434,7 +1434,7 @@ func CompareAndPullRequestPost(ctx *context.Context) {
 				return
 			}
 			ctx.Flash.Error(flashError)
```
Comment (yp05327) [inline routers/web/repo/pull.go]:
```suggestion
```
This is unnecessary.

**Comment by CaiCandong** [comment]
I haven't figured out how this Flash works.，why I only get network errors with the following function？@zeripath
```			
ctx.Flash.Error(flashError, true)
```

**Comment by yp05327** [comment]
I think we can not use Flash error here, we should use `JSONError` and handling the msg by js.
But we are using tippy to show this error msg which can not render html code correctly, 
so we need do more changes as the comment said here:
https://github.com/go-gitea/gitea/blob/a370efc13f0e1ea309e324639832832bc14cb6dc/web_src/js/features/common-global.js#L122

**Comment by silverwind** [comment]
The toast module may be more suitable, but it currently also escapes HTML for security reason. If we are sure there is nothing user-supplied in these messages, we could add an option to disable HTML escaping. Same is also possible for Tippy, but generally I'm wary because the screenshot above does show it contains user data.

https://github.com/go-gitea/gitea/blob/main/web_src/js/modules/toast.js

**Comment by yp05327** [comment]
Maybe we can remove `Details` in the error message?
Then using toast will not cause security problems.

ps: Did #26259 fully fix #26129? #26129 is closed now.

**Pull request closed without merging**